Added rio extract command

brendan-ward opened the pull request:
Resolves #243 

Adds an ```extract``` command to rio which allows you to extract out pixels from a raster using features to either mask in (default) or mask out (using ```--invert```) pixels.

Includes minor refactoring of window handling to consolidate usages.  Note: this is an API change, do to the way default value is set.  Only one test needed to be updated, but this should be noted in the changelog if this PR is accepted.

brendan-ward commented on rasterio/_base.pyx:
```diff
@@ -373,11 +373,15 @@ cdef class DatasetReader(object):
-        ul = self.index(left, top)
-        lr = self.index(right, bottom)
-        return tuple(zip(ul, lr))
+    def window(self, left, bottom, right, top, boundless=False):
```
@sgillies here is the API change, using ```boundless``` as you suggested.

sgillies replied:
:+1:

sgillies commented on rasterio/features.py:
```diff
@@ -205,11 +248,11 @@ def rasterize(
         If True, all pixels touched by geometries will be burned in.  If
         false, only pixels whose center is within the polygon or that
-        are selected by brezenhams line algorithm will be burned in.
+        are selected by Brezenhams line algorithm will be burned in.
```
Bresenham.

sgillies commented on rasterio/rio/features.py:
```diff
@@ -20,6 +19,128 @@
+@cli.command(short_help='Extract raster using features.')
+@files_inout_arg
+@format_opt
+@click.option('--all_touched', is_flag=True, default=False,
```
`--all-touched`.

brendan-ward replied:
Current usage matches rasterize.  Do you want that changed too?  We should be consistent...

sgillies replied:
Yes, let's change rasterize, too. Classic unix style options with dashes as word separators is our style.

sgillies replied:
Oh, crap. We're not consistent about this at all and I seem to be the worst offender (cligj as example).

brendan-ward replied:
Ok - I'll fix both, then do another sweep later per #324

sgillies commented on rasterio/rio/features.py:
```diff
@@ -20,6 +19,128 @@
+
+    --crop option is not valid if features are completely outside extent of
+    input raster.
+    """
```
@brendan-ward This program strikes me as being a filter like `grep`. What would you think about borrowing command line options from grep? `--invert` could become `-v, --invert-mask` (like grep's `-v, --invert-match`). And I'd like to see the GeoJSON be more like grep's `pattern`: not an input filename but either a GeoJSON text argument or an `-f/--file features.json` option. If there were no features, we could pass all the input pixels through.

So, the canonical example might be 

```
$ rio extract -f features.json input.tif output.tif
```



```
     
             Selected lines are those not matching any of the specified pat-
             terns.
```

brendan-ward replied:
@sgillies Interesting idea, I'll have to think on it...

I do worry about use of ```-v``` here, since you are using it for verbosity [elsewhere](https://github.com/mapbox/rasterio/blob/master/rasterio/rio/info.py#L71), and ```-f``` is used here and in cligj for [output format](https://github.com/mapbox/cligj/blob/master/cligj/__init__.py#L38) 

I was trying to keep the usage pretty similar to ```rasterize``` (in terms of stdin or filename) so that it felt familiar to folks, and thinking there would be use cases where we pipe in GeoJSON rather than have it in advance, and this usage looks a little [awkward in grep](http://stackoverflow.com/questions/6641731/read-expression-for-grep-from-standard-input).

What do you expect the output of bounding box extraction to be, as specified above?   Original extent of input.tif, but masked out everywhere except [west, south, east, north]?  Or cropped to that extent as well?  When I was considering whether or not I should add this, it looks like gdal_translate covers the crop to extent of bbox pretty well, but doesn't combine that with masking (but result would be the same if we are only specifying a rectangular bbox in world coordinates of raster).

sgillies replied:
How about `-i/--invert` and `-j/--geojson-file`?

Yes, I see the difference in between rio-rasterize and what I'm suggesting. Piping in GeoJSON would be nice, I agree. I'm concerned it interferes with being able to pipe or pipeline rasters (#188), which is what we should be aiming for in the suite of rio commands. In the rio-rasterize case, there aren't any raster inputs, so it's fine to turn everything over to GeoJSON and have it be the foremost input.

I'm also a tiny bit concerned about confusion around the term "extract". We already refer to feature extraction in a bunch of places (like the rio-shapes docs). Could this be better called "rio-select" or "rio-mask"?

dnomadb commented:
> Piping in GeoJSON would be nice, I agree

This would be very useful for us, i think. Especially in terms of being able to couple libraries like `mercantile` (example usage: `mercantile shapes <tile> | rio extract`).

That said, being able to extract / clip from a simple '[w, s, e, n]' bounding box would also be great.

> I'm also a tiny bit concerned about confusion around the term "extract"

What about rio clip? This seems like the most conspicuous name for this functionality.

brendan-ward commented:
Use of extract was based on my background with similar tools in [ArcGIS: Extract by Polygon](http://help.arcgis.com/en/arcgisdesktop/10.0/help/index.html#//009z0000002q000000.htm).

I'm definitely open to changing the command name to make it more clear.  I think ```rio mask``` may be a good fit - the primary operation is masking.  That said, do we see a case for allowing masking by another raster - e.g., you give us a masking file that is either GeoJson and we rasterize it, or you give us a raster mask (in same CRS, and we deal w/ shape & transform properly) where we mask out everything in the input raster except pixels covered by a non-nodata value in the mask raster?

I like the idea of pipelining rasters, so I'll set about refactoring this to better support that.

Maybe we should add ```rio clip``` as a separate command, and keep the options / scope of that one really tight?  That is my preference at the moment, because it doesn't depend on masking...

brendan-ward commented:
OK - I renamed to ```rio mask```, updated the options per #324 and added short versions for ```mask``` and ```rasterize```, made GeoJSON an optional input.

I've done everything except adding clip from a bounding box or ```rio clip```.  Any consensus on whether that functionality belongs here or in a new command?

sgillies commented:
@brendan-ward I was out sick today, will catch up on this in the morning.

sgillies commented:
@brendan-ward let's move ahead in a few phases:

1. yank short names we're uncertain of
2. merge this PR
3. work on short names and consolidation of opts in a new branch

brendan-ward commented:
@sgillies pending Travis, this should now be ready to go.  I think the [bike shed](http://en.wikipedia.org/wiki/Parkinson's_law_of_triviality) would look great in [slate teal](http://www.benjaminmoore.com/en-us/paint-color/slateteal), but as for the trim... :)

perrygeo commented:
@brendan-ward I'd like to revisit the bounding box clipping you mentioned. Was there a reason this was excluded? 

I think a separate `rio clip` command would be appropriate since it matches the common terminology and doesn't actually involve masking. Accepting a bbox directly ala `fio bounds` would be ideal.

```
fio dump data.shp | fio bounds | rio-clip in.tif out.tif
```
There's already a [rio-clip](https://github.com/kapadia/rio-clip) rasterio plugin though which complicates naming. 

An alternative would just be to use the `rio mask` command as is but modify `fio bounds` to output GeoJSON Polygons. This sort of thing is currently possible at the command line but you've got to jump through a lot of hoops...

```shell
fio dump ../austria.shp \
  | fio bounds \
  | turf bboxPolygon "$(cat /dev/stdin)" \
  | rio mask --crop -j - srtm.tif srtm_austria.tif
```

brendan-ward commented:
It got omitted more from my forgetfulness than by intent; as you can see there was much more back and forth about naming of flags than around the clip functionality and I didn't loop back on this.  Sorry!

I still like the idea of a separate command to keep the functionality tightly bounded.

What about the idea of hybridizing this approach w/ plugin from @kapadia so that we have one of a couple pathways

```rio clip in.tif out.tif --bounds [w, s, e, n]```  
where bounds could default to stdin so you could do it as you specified above:
```fio dump data.shp | fio bounds | rio clip in.tif out.tif```

or (from @kapadia's plugin):

```rio clip in.tif out.tif --like template.tif```


@kapadia what say you - can we pull your logic in here instead of needing a plugin?


Once we've got the API set I'd be happy to take this on.

\cc @dnomadb @sgillies

kapadia commented:
@brendan-ward @perrygeo Sure, no problem in pulling in `rio-clip` logic into rasterio. I've stopped using the plugin since we enabled `--dst-crs` for `rio bounds`. These days I do the following:

    rio bounds in1.tif --dst-crs EPSG:XXXX | rio mask in2.tif out.tif --geojson-mask - --crop

If I follow correctly, the only difference in @perrygeo's case is working with a vector file. A `rio-clip` function that is aware of bounding boxes would be nice (especially coupled with [`mercantile`](https://github.com/mapbox/mercantile)). GeoJSON output in Fiona coupled with `rio-mask` would also provide nice solution.

brendan-ward commented:
@kapadia what are you seeing is needed to make rio-clip as defined above work with mercantile?  An example of the command line call that couples the two together would prove helpful to me in understanding what you mean.

I wasn't thinking of reprojection when I suggested ```--like``` option above.  But we could easily handle that, since we know source and destination CRS with more confidence than we do for GeoJSON inputs...

kapadia commented:
Ah! I had forgotten the geojson output of `mercantile`. I was thinking of the bbox output:

    $ echo "[17, 13, 5]" | mercantile shapes --bbox
    [11.25, 21.943045533438177, 22.5, 31.952162238024968]

Anyway, I wrote `rio-clip` to resolve projection differences when clipping an image to the bounds of another. Often I'm working with images in a slightly different projection. If you recall, adding `--dst-crs` to `rio-bounds` resolved my use case, and I used the existing `rio-mask`.

perrygeo commented:
If we could rely on `rio bounds --bbox`, `fio bounds`, and `mercantile shape --bbox` to translate everything into simple bbox array strings, The `rio clip` API becomes very simple:

```
Usage: rio clip [OPTIONS] [BBOX]

Options:
  --input PATH         Path to input raster.
  -o, --output PATH  Path to output raster.
```
How does this sound @brendan-ward @kapadia  @sgillies ? 

I've got time this afternoon to start a PR if @brendan-ward  hasn't started on it already...

sgillies commented:
@perrygeo another option: extending rio-convert with a `--bbox` option.

But if we kept this as a separate function, I think the right usage would be 

```
Usage: rio clip [OPTIONS] INPUT BBOX OUTPUT

Options:
  -o, --output PATH               Path to output file (optional alternative to
                                  a positional arg for some commands).
```

See https://github.com/mapbox/rasterio/blob/master/rasterio/rio/convert.py#L18-L34 for rio-convert's implementation.

brendan-ward commented:
I've gotten most of the way through the implementation as a separate function, still need to do tests.

I'm struggling to balance use of ```--bounds W S N E``` (4 param form) as used for other functions versus an additional option here to allow a JSON array string that can also be read from ```stdin```.  (right now ```--bounds_json``` but could easily be called ```--bbox```, which is better, or something else).

Do we need or want both variants here?

I've also got an option for using a ```--like``` raster as a source of bounds, reprojected as necessary.  Do we not want that here?

perrygeo commented:
I think the `--like` option would be unnecessary since we can grab the bounds of any raster with `rio bounds --bbox` and pipe it in

Re: the bounds issue, I think we could probably accept a variety of formats (comma delimited, space delimited, json array) so long as they were one argument and in WSNE order. I'd definitely vote for a stdin rather than a option.

perrygeo commented:
@sgillies +1 to that usage example. 

`convert` would also be a good fit technically, but `clip` is semantically more accurate for the task at hand  (years of Desktop GIS ©  use has burned that vocabulary into my brain)

brendan-ward commented:
I like where we are heading in terms of simplicity.  Will try to get a PR in tonight.

sgillies commented:
Yes, supporting `[w, s, e, n]` or `w, s, e, n` or `w s e n` isn't too complicated and is a usability win. I'm doing that type of thing in https://github.com/mapbox/mapbox-sdk-py/issues/8.
